packages/openmrs/src/Utils.js

```diff
@@ -51,43 +62,84 @@ export async function request(state, method, path, options = {}) {
   };
 
   const url = `${baseUrl}${path}`;
+  let queryParams = opts?.query;
+  const requestOptions = queryParams ? { ...opts, query: queryParams } : opts;
+
+  return (shouldPaginate({ underGlobalLimit: true, hasPageSize: query.pageSize }))
+    ? paginate(method, url, requestOptions)
+    : fetchAndLog(method, url, requestOptions);
+}
+
+export async function paginate(method, url, opts) {
 
   let allResponses;
-  let queryParams = opts?.query;
-  let allowPagination = isNaN(queryParams?.startIndex);
+  let fetchedCount = 0;
+
+
+  // global limit variable that is a source of truth incase there is a pageSize variable
+  const globalLimit = opts.query.limit;
+
+  // Declare a variable that takes in all the options. We can then modify this variable to fetch the next page
+  let paginatedRequestOptions = { ...opts }
+
+
+  // if page size variable exists use that as limit. Delete pageSize from params since OpenMRS does recognize it.
+  if (opts.query.pageSize) {
+    paginatedRequestOptions.query.limit = opts.query.pageSize;
```

Comment:
This needs to be pagesize OR (total items - current items), whichever is smaller